File: week 4/shortest_paths/shortest_paths.py
#Uses python3
#import sys
#import queue

#Uses python2
import Queue


def shortet_paths(adj, cost, s, distance, reachable, shortest):
    #write your code here
    distance[s] = 0
    reachable[s] = 1
    q = Queue.Queue()
    for i in range(len(adj)):
        for u in range(len(adj)):
            for v in adj[u]:
                v_index = adj[u].index(v)
                if distance[u] != 10**19 and distance[v] > distance[u] + cost[u][v_index]:
                    distance[v] = distance[u] + cost[u][v_index]
                    reachable[v] = 1
                    if i == len(adj) - 1:
                        q.put(v)

    visited = [0] * len(adj)
    while not q.empty():
        u = q.get()
        visited[u] = 1
        if u != s:
            shortest[u] = 0
        for v in adj[u]:
            if visited[v] == 0:
                q.put(v)
                visited[v] = 1
                shortest[v] = 0
    distance[s] = 0

if __name__ == '__main__':
    n, m = map(int, raw_input().split())
    adj = [[] for _ in range(n)]
    cost = [[] for _ in range(n)]
    for i in range(m):
        a, b, w = map(int, raw_input().split())
        adj[a - 1].append(b - 1)
        cost[a - 1].append(w)
    s = int(raw_input())
    '''
    input = sys.stdin.read()
    data = list(map(int, input.split()))
    n, m = data[0:2]
    data = data[2:]
    edges = list(zip(zip(data[0:(3 * m):3], data[1:(3 * m):3]), data[2:(3 * m):3]))
    data = data[3 * m:]
    adj = [[] for _ in range(n)]
    cost = [[] for _ in range(n)]
    for ((a, b), w) in edges:
        adj[a - 1].append(b - 1)
        cost[a - 1].append(w)
    s = data[0]
    '''
    s -= 1
    distance = [10**19] * n
    reachable = [0] * n
    shortest = [1] * n
    shortet_paths(adj, cost, s, distance, reachable, shortest)
    for x in range(n):
        if reachable[x] == 0:
            print('*')
        elif shortest[x] == 0:
            print('-')
        else:
            print(distance[x])

File: week 4/shortest_paths/shortest_paths.cpp
#include <iostream>
#include <limits>
#include <vector>
#include <queue>

using std::vector;
using std::queue;
using std::pair;
using std::priority_queue;

void shortest_paths(vector<vector<int> > &adj, vector<vector<int> > &cost, int s, vector<long long> &distance, vector<int> &reachable, vector<int> &shortest) {
  //write your code here
  reachable[s] = 1;
  distance[s] = 0;
  queue<int> q;
  for (int i = 0; i < adj.size(); i++) {
    for (int u = 0; u < adj.size(); u++){
      for (int k = 0; k < adj[u].size(); k++) {
        int v = adj[u][k];
        if (distance[u]!= std::numeric_limits<long long>::max() && distance[v] > distance[u] + cost[u][k]) {
          distance[v] = distance[u] + cost[u][k];
          reachable[v] = 1;
          if (i == adj.size() - 1) {
            q.push(v);
          }	
        }
      }
    }
  }
  
  vector<int> visited(adj.size());
  while (!q.empty()){
    int u = q.front();
	q.pop();
	visited[u] = 1;
// 	if(u != s)
	shortest[u] = 0;
	for (int i = 0; i < adj[u].size(); i++) {
	  int v = adj[u][i];
      if (!visited[v]) {
        q.push(v);
        visited[v] = 1;
        shortest[v] = 0;
	  }
	}
  }
  distance[s] = 0;
}

int main() {
  int n, m, s;
  std::cin >> n >> m;
  vector<vector<int> > adj(n, vector<int>());
  vector<vector<int> > cost(n, vector<int>());
  for (int i = 0; i < m; i++) {
    int x, y, w;
    std::cin >> x >> y >> w;
    adj[x - 1].push_back(y - 1);
    cost[x - 1].push_back(w);
  }
  std::cin >> s;
  s--;
  vector<long long> distance(n, std::numeric_limits<long long>::max());
  vector<int> reachable(n, 0);
  vector<int> shortest(n, 1);
  shortest_paths(adj, cost, s, distance, reachable, shortest);
  for (int i = 0; i < n; i++) {
    if (!reachable[i]) {
      std::cout << "*\n";
    } else if (!shortest[i]) {
      std::cout << "-\n";
    } else {
      std::cout << distance[i] << "\n";
    }
  }
}
